### DISCUSSION TITLE
Plan: Refactor `[params]` Config — snake_case + Flatten `params.page.xxx`

### DISCUSSION THREAD

**Discussion by Lruihao** [hugo-fixit/FixIt#796] [Dev]
## Context

The `hugo.toml` params config has two problems:
1. **Inconsistent naming**: mix of camelCase (`defaultTheme`, `hiddenFromHomePage`), snake_case (`short_name`, `global_font_family`), and PascalCase (`GitHub`, `Twitter`)
2. **`[params.page]` wrapper layer** prevents using Hugo's built-in `.Param` — the theme had to invent `function/params.html` with `page.Params | merge site.Params.page` as a workaround

The existing `snake2camel.html` partial (line 29) already documents this as a planned v1.0 breaking change.

## Goals

1. **Eliminate `[params.page]`** — move ALL keys (flat + sub-sections) to `[params]` level
2. **Remove `function/params.html`** — no more manual merge, use Hugo's `.Param` throughout
3. All params keys → **snake_case** in `hugo.toml`
4. Templates and JS keep camelCase — `snake2camel.html` handles conversion where needed

## Not Changed

- **PascalCase social platform names** in `[params.social]` and `[params.share]` — identifiers, not config keys
- **SCSS variable names** in `[params.appearance]` — already snake_case
- **HTML data-attribute keys** in analytics — already snake_case
- **TypeScript interfaces** in `assets/js/types/config.ts` — stay camelCase

---

## Step 1: `hugo.toml` — Eliminate `[params.page]`, move everything to `[params]`

All keys under `[params.page]` move to `[params]`. The `[params.page]` section is deleted entirely.

**Flat keys** (currently `[params.page]`):
`authorAvatar`, `hiddenFromHomePage`, `hiddenFromSearch`, `hiddenFromRelated`, `hiddenFromFeed`, `twemoji`, `lightgallery`, `ruby`, `fraction`, `fontawesome`, `license`, `linkToMarkdown`, `linkToSource`, `linkToEdit`, `linkToReport`, `linkToVscode`, `pageStyle`, `autoBookmark`, `showLastmod`, `wordCount`, `readingTime`, `endFlag`, `instantPage`, `collectionList`, `collectionNavigation`, `password`, `featuredImage`, `featuredImagePreview`, `subtitle`, `editURL`, `sourceURL`, `weight`, `message`

**Sub-sections** (currently `[params.page.xxx]`):
`toc`, `math` (+ katex, mathjax), `comment` (+ all comment systems), `share`, `reward` (+ ways), `related`, `mapbox`, `heading` (+ number, format), `expirationReminder`, `repost`, `library` (+ css, js), `seo` (+ publisher)

## Step 2: `hugo.toml` — Convert all camelCase keys to snake_case

**Top-level `[params]`**: `defaultTheme` → `default_theme`, `dateFormat` → `date_format`, `capitalizeTitles` → `capitalize_titles`, `withSiteTitle` → `with_site_title`, `disableThemeInject` → `disable_theme_inject`, etc.

**Section names to rename**:
- `[params.gitInfo]` → `[params.git_info]`
- `[params.githubCorner]` → `[params.github_corner]`
- `[params.backToTop]` → `[params.back_to_top]`
- `[params.readingProgress]` → `[params.reading_progress]`
- `[params.repoVersion]` → `[params.repo_version]`
- `[params.taskList]` → `[params.task_list]`
- `[params.postChat]` → `[params.post_chat]`
- `[params.postSummary]` → `[params.post_summary]`
- `[params.jsonViewer]` → `[params.json_viewer]`
- `[params.customPartials]` → `[params.custom_partials]`
- `[params.recentlyUpdated]` → `[params.recently_updated]`

**`[params.page]` flat keys → `[params]` with snake_case**:
`author_avatar`, `hidden_from_home_page`, `hidden_from_search`, `hidden_from_related`, `hidden_from_feed`, `link_to_markdown`, `link_to_source`, `link_to_edit`, `link_to_report`, `link_to_vscode`, `page_style`, `auto_bookmark`, `show_lastmod`, `word_count`, `reading_time`, `end_flag`, `instant_page`, `collection_list`, `collection_navigation`, `featured_image`, `featured_image_preview`

**Migrated sub-section internal keys**:
- `[params.toc]`: `keepStatic` → `keep_static`, `startLevel` → `start_level`, `endLevel` → `end_level`, `decreaseH1` → `decrease_h1`
- `[params.math.katex]`: `copyTex` → `copy_tex`, `throwOnError` → `throw_on_error`, `errorColor` → `error_color`
- `[params.math.mathjax.options]`: `enableMenu` → `enable_menu`, `skipHtmlTags` → `skip_html_tags`, `ignoreHtmlClass` → `ignore_html_class`
- `[params.mapbox]`: `accessToken` → `access_token`, `lightStyle` → `light_style`, `darkStyle` → `dark_style`
- `[params.heading.number]`: `onlyMainSection` → `only_main_section`
- `[params.expiration_reminder]`: `closeComment` → `close_comment`
- `[params.seo.publisher]`: `logoUrl` → `logo_url`
- Comment systems: `clientId` → `client_id`, `serverURL` → `server_url`, `appId` → `app_id`, `pageSize` → `page_size`, `commentCount` → `comment_count`, etc.
- Plus all other camelCase keys in search, header, footer, watermark, mermaid, typeit, busuanzi, etc.

## Step 3: Remove `function/params.html` and use `.Param`

**Delete**: `layouts/_partials/function/params.html`

**Hugo's `.Param` behavior**: searches page frontmatter first, falls back to `site.Params`, case-insensitive. Works for both scalar and map values.

### 3.1 Replace `$params.xxx` with `.Param "xxx"`

Every template that currently does:
```go-html-template
{{- $params := partial "function/params.html" -}}
... $params.toc ... $params.ruby ... $params.password ...
```

Becomes direct `.Param` calls where the value is used:
```go-html-template
.Param "toc" ... .Param "ruby" ... .Param "password" ...
```

For map values (sub-sections), `.Param "toc"` returns the entire map — same as `$params.toc` did.

### 3.2 Replace inline merge patterns

Templates with `.Params | merge .Site.Params.page` or `.Page.Params | merge site.Params.page` simply use `.Param "xxx"` or `.Page.Param "xxx"` directly — no merge needed.

### 3.3 Key-by-key conversion table (snake_case names)

| Old `$params.xxx` | New `.Param "xxx"` | Type |
|---|---|---|
| `$params.ruby` | `.Param "ruby"` | scalar |
| `$params.fraction` | `.Param "fraction"` | scalar |
| `$params.fontawesome` | `.Param "fontawesome"` | scalar |
| `$params.password` | `.Param "password"` | scalar |
| `$params.lightgallery` | `.Param "lightgallery"` | scalar |
| `$params.twemoji` | `.Param "twemoji"` | scalar |
| `$params.subtitle` | `.Param "subtitle"` | scalar |
| `$params.auto_bookmark` | `.Param "auto_bookmark"` | scalar |
| `$params.instant_page` | `.Param "instant_page"` | scalar |
| `$params.show_lastmod` | `.Param "show_lastmod"` | scalar |
| `$params.word_count` | `.Param "word_count"` | scalar |
| `$params.reading_time` | `.Param "reading_time"` | scalar |
| `$params.end_flag` | `.Param "end_flag"` | scalar |
| `$params.author_avatar` | `.Param "author_avatar"` | scalar |
| `$params.collection_list` | `.Param "collection_list"` | scalar |
| `$params.collection_navigation` | `.Param "collection_navigation"` | scalar |
| `$params.hidden_from_search` | `.Param "hidden_from_search"` | scalar |
| `$params.hidden_from_home_page` | `.Param "hidden_from_home_page"` | scalar |
| `$params.hidden_from_related` | `.Param "hidden_from_related"` | scalar |
| `$params.license` | `.Param "license"` | scalar |
| `$params.link_to_markdown` | `.Param "link_to_markdown"` | scalar |
| `$params.link_to_source` | `.Param "link_to_source"` | scalar |
| `$params.link_to_edit` | `.Param "link_to_edit"` | scalar |
| `$params.link_to_report` | `.Param "link_to_report"` | scalar |
| `$params.link_to_vscode` | `.Param "link_to_vscode"` | scalar |
| `$params.page_style` | `.Param "page_style"` | scalar |
| `$params.featured_image` | `.Param "featured_image"` | scalar |
| `$params.featured_image_preview` | `.Param "featured_image_preview"` | scalar |
| `$params.edit_url` | `.Param "edit_url"` | scalar |
| `$params.source_url` | `.Param "source_url"` | scalar |
| `$params.message` | `.Param "message"` | scalar |
| `$params.weight` | `.Param "weight"` | scalar |
| `$params.toc` | `.Param "toc"` | map |
| `$params.math` | `.Param "math"` | map |
| `$params.comment` | `.Param "comment"` | map |
| `$params.share` | `.Param "share"` | map |
| `$params.reward` | `.Param "reward"` | map |
| `$params.related` | `.Param "related"` | map |
| `$params.mapbox` | `.Param "mapbox"` | map |
| `$params.heading` | `.Param "heading"` | map |
| `$params.expiration_reminder` | `.Param "expiration_reminder"` | map |
| `$params.repost` | `.Param "repost"` | map |
| `$params.library` | `.Param "library"` | map |
| `$params.seo` | `.Param "seo"` | map |

## Step 4: Update `init/patch.html` — shorthand expansion

Keep shorthand boolean handling, but read site defaults from `site.Params.xxx` instead of `site.Params.page.xxx`:

```go-html-template
{{- /* Toc shorthand patch */ -}}
{{- $toc := .Param "toc" -}}
{{- if not (reflect.IsMap $toc) -}}
  {{- $toc = dict "enable" $toc | merge site.Params.toc -}}
{{- end -}}
{{- .Store.Set "toc" $toc -}}

{{- /* Reward shorthand patch */ -}}
{{- $reward := .Param "reward" -}}
{{- if eq $reward true -}}
  {{- $reward = dict "enable" true | merge site.Params.reward -}}
{{- else if eq $reward false -}}
  {{- $reward = dict "enable" false -}}
{{- end -}}
{{- .Store.Set "reward" $reward -}}

{{- /* Math shorthand patch */ -}}
{{- $math := .Param "math" -}}
{{- if eq $math true -}}
  {{- $math = dict "enable" true | merge site.Params.math -}}
{{- else if eq $math false -}}
  {{- $math = dict "enable" false -}}
{{- end -}}
{{- .Store.Set "math" $math -}}
```

## Step 5: Update `init/detection-env.html` — comment shorthand

```go-html-template
{{- .Store.Set "comment" .Param "comment" -}}
{{- if eq .Params.comment true -}}
  {{- .Store.Set "comment" site.Params.comment -}}
{{- else if eq .Params.comment false -}}
  {{- .Store.Set "comment" (dict "enable" false) -}}
{{- end -}}
```

### 6.1 Files calling `partial "function/params.html"` → remove, use `.Param`

| File | Change |
|---|---|
| `layouts/page.html` | Remove `$params`, use `.Param "ruby"` etc. |
| `layouts/home.html` | Remove `$params`, use `.Param "ruby"` etc. |
| `layouts/search.html` | Remove `$params`, use `.Param` |
| `layouts/friends.html` | Remove `$params`, use `.Param` |
| `layouts/posts/single.html` | Remove `$params`, use `.Param` for ~15 keys |
| `layouts/_partials/init/patch.html` | Remove `$params`, use `.Param "toc"` etc. |
| `layouts/_partials/init/detection-env.html` | Remove `$params`, use `.Param "comment"` |
| `layouts/_partials/plugin/pagefind-metadata.html` | Remove `$params`, use `.Param` |
| `layouts/_partials/plugin/share.html` | `(partial).share` → `.Param "share"` |
| `layouts/_partials/single/footer.html` | Remove `$params`, use `.Param` for ~10 keys |
| `layouts/_partials/single/expiration-reminder.html` | Remove `$params`, use `.Param` |
| `layouts/_partials/single/collection-list.html` | Remove `$params`, use `.Param` |
| `layouts/_partials/single/collection-nav.html` | Remove `$params`, use `.Param` |
| `layouts/_partials/single/related-aside.html` | Remove `$params`, use `.Param` |
| `layouts/_partials/single/related.html` | Remove `$params`, use `.Param` |
| `layouts/_partials/base/widgets.html` | Remove `$params`, use `.Param` |
| `layouts/_partials/base/head/index.html` | Remove `$params`, use `.Param "seo"` |
| `layouts/_partials/base/assets.html` | Remove `$params`, use `.Param` for ~9 keys |
| `layouts/_partials/base/comment.html` | Remove `$params`, use `.Param "lightgallery"` |

### 6.2 Inline merge patterns → remove, use `.Param`

| File | Change |
|---|---|
| `layouts/summary.html` | Remove merge, use `.Param` |
| `layouts/_partials/single/post-author.html` | Remove merge, use `.Param "author_avatar"` |
| `layouts/_markup/render-image.html` | Remove merge, use `.Page.Param "lightgallery"` |
| `layouts/_markup/render-passthrough.html` | Remove merge, use `.Page.Param "math"` |
| `layouts/_markup/render-heading.html` | Remove merge, use `.Page.Param "toc"` / `.Page.Param "heading"` |
| `layouts/_shortcodes/mapbox.html` | Remove merge, use `.Page.Param "mapbox"` |
| `layouts/_partials/feed/rss.html` | Remove merge, use `.Param` |
| `layouts/_partials/plugin/post-chat-ai.html` | Already merges with `.Site.Params`; use `.Param` |

### 6.3 Direct `.Site.Params.page.xxx` → `.Param "xxx"` or `site.Params.xxx`

| File | Old | New |
|---|---|---|
| `layouts/home.html:17` | `.Site.Params.page.hiddenFromHomePage` | `.Param "hidden_from_home_page"` (in `where` filter: `"Param \"hidden_from_home_page\""`) |
| `layouts/_partials/init/patch.html:7,14,23` | `.Site.Params.page.toc/reward/math` | `site.Params.toc/reward/math` |
| `layouts/_partials/init/detection-env.html:19` | `.Site.Params.page.comment` | `site.Params.comment` |
| `layouts/_partials/single/related.html:7` | `.Site.Params.page.hiddenFromRelated` | `.Param "hidden_from_related"` |
| `layouts/_partials/single/related-aside.html:7` | `.Site.Params.page.hiddenFromRelated` | `.Param "hidden_from_related"` |
| `layouts/_partials/plugin/post-chat-ai.html:20` | `.Site.Params.page.password` | `site.Params.password` (fallback) |

### 6.4 `where` filter updates

- `Params.hiddenfromhomepage` → `Param "hidden_from_home_page"` (`.Param` is case-insensitive)
- `Params.hiddenfromfeed` → `Param "hidden_from_feed"`
- `Params.hiddenfromrelated` → `Param "hidden_from_related"`
- `Params.hiddenfromsearch` → `Param "hidden_from_search"`

### 6.5 `.Site.Params.xxx` renamed key updates

All `.Site.Params.camelCaseKey` references in templates must use the new snake_case names. Hugo's `.Param` is case-insensitive, but `.Site.Params.xxx` map access IS case-sensitive. Every grep hit must be updated.

Key files: `assets.html` (~50 references), `comment.html`, `widgets.html`, `head/index.html`, `head/css.html`, `rss.html`, `post-chat-ai.html`, `recently-updated.html`, `task-lists.html`, etc.

## Step 7: Update `window.config` construction in `assets.html`

The `$config` dict keys emitted to JS stay camelCase (JS-facing). Only source value access changes to read from snake_case config keys:
- `$typeit.cursorSpeed` → `$typeit.cursor_speed`
- `$watermark.rowSpacing` → `$watermark.row_spacing`
- etc.

## Step 8: Deprecation warnings in `init/detection-deprecated.html`

Update `detection-deprecated.html` to detect old config keys and emit `warnf` messages. Hugo's `site.Params` is read-only, so we cannot do runtime migration — users must update their config manually.

### 8.1 `params.page.*` structural migration warnings

Check if old `params.page` sub-sections exist and warn:

```go-html-template
{{- with .Site.Params.page -}}
  {{- if reflect.IsMap . -}}
    {{- range $key, $value := . -}}
      {{- $warns = $warns | append (printf "The parameter `params.page.%s` is deprecated, use `params.%s` instead." $key $key) -}}
    {{- end -}}
  {{- end -}}
{{- end -}}
```

This catches ALL keys that users still have under `[params.page]`, both flat keys and sub-sections.

### 8.2 camelCase key warnings

For commonly-used camelCase keys, add explicit warnings. Hugo's `.Param` is case-insensitive so frontmatter keys still work, but site config keys must be renamed:

```go-html-template
{{- with .Site.Params.defaultTheme -}}
  {{- $warns = $warns | append "The parameter `params.defaultTheme` is deprecated, use `params.default_theme` instead." -}}
{{- end -}}
{{- with .Site.Params.dateFormat -}}
  {{- $warns = $warns | append "The parameter `params.dateFormat` is deprecated, use `params.date_format` instead." -}}
{{- end -}}
{{- with .Site.Params.capitalizeTitles -}}
  {{- $warns = $warns | append "The parameter `params.capitalizeTitles` is deprecated, use `params.capitalize_titles` instead." -}}
{{- end -}}
{{- with .Site.Params.withSiteTitle -}}
  {{- $warns = $warns | append "The parameter `params.withSiteTitle` is deprecated, use `params.with_site_title` instead." -}}
{{- end -}}
{{- with .Site.Params.disableThemeInject -}}
  {{- $warns = $warns | append "The parameter `params.disableThemeInject` is deprecated, use `params.disable_theme_inject` instead." -}}
{{- end -}}
```

### 8.3 Backward compatibility note

Hugo's `site.Params` is read-only — we cannot copy old keys to new locations at runtime. The structural change (`params.page.xxx` → `params.xxx`) is a hard breaking change for v1.0. Users must update their `hugo.toml` manually.

However, `.Param` is case-insensitive, so the snake_case rename alone is NOT a breaking change for frontmatter — old camelCase frontmatter keys (e.g., `hiddenFromHomePage: true`) still work via `.Param "hidden_from_home_page"`.

## Step 9: Delete `function/params.html`

Remove the file once all references are converted.

---

## Critical Files

| File | Changes |
|---|---|
| `hugo.toml` | All structural moves + key renames |
| `layouts/_partials/function/params.html` | **DELETE** |
| `layouts/_partials/init/patch.html` | Use `.Param`, update `site.Params.xxx` paths |
| `layouts/_partials/init/detection-env.html` | Use `.Param`, update `site.Params.xxx` path |
| `layouts/_partials/base/assets.html` | ~50 param references → `.Param` + snake_case |
| `layouts/_partials/base/comment.html` | Comment config key access |
| `layouts/_partials/base/widgets.html` | backToTop, githubCorner, readingProgress |
| `layouts/posts/single.html` | ~15 `$params` → `.Param` calls |
| `layouts/summary.html` | Remove merge, use `.Param` |
| `layouts/_markup/render-*.html` | Remove merge, use `.Page.Param` |
| `layouts/_partials/init/detection-deprecated.html` | Add new deprecation warnings |

## Verification

1. `pnpm build:demo` — verify demo site builds without errors
2. `pnpm build:test` — verify test site builds without errors
3. `pnpm lint` — ESLint passes
4. `pnpm typecheck` — TypeScript passes
5. Manually inspect built demo site: TOC, math, comments, search, share, lightgallery, mapbox all work
